CIP-016X/README.md

```diff
@@ -0,0 +1,613 @@
+3. **Attack Pattern Detection**: Identify systematic invalid block patterns that may indicate coordinated attacks
+4. **Peer Attribution**: Track which peers send invalid blocks to build reputation profiles
+5. **False Positive Filtering**: Distinguish between genuine attacks and benign misconfigurations
+
+**Expensive-to-Validate Detection**:
+
+Background workers measure validation duration to identify blocks and transactions that are computationally expensive to process. This helps detect potential DoS vectors, analyze SPO behavior, and monitor network health.
+
+**Measurement Approach**:
+
+1. **Timing Capture**: Record start and end timestamps when background workers submit blocks/transactions to the full node for validation
+2. **Duration Calculation**: Compute validation duration as the round-trip time of the validation query
+3. **Threshold Detection**: Flag blocks/transactions exceeding the configured `expensive_validation_threshold_seconds`
+4. **Metadata Storage**: Store validation timing data alongside block/transaction records via Repository operations
+
+**Limitations**:
+
+- **Best-effort metric**: Validation timing includes network latency and full node processing queue delays, not just pure validation cost
+- **Caching effects**: Full node may cache validation results; repeated validation of the same block returns cached results instantly
+- **Cold vs warm validation**: First-time validation may be slower due to UTxO cache warming and ledger state access patterns
+- **Concurrent load**: Full node load from other operations affects measured validation times
+
+**Validation Cost Factors**:
+
+Blocks and transactions vary significantly in validation cost due to factors such as Plutus script complexity, transaction count, and UTxO operations. For detailed analysis of validation cost factors and measurement approaches, see the [Block Cost Investigation proposal](https://docs.google.com/document/d/1mDA0NuD7rQYtNMmwY2zNz9wci7RCy2bMPFZrbolA-jA/edit?tab=t.0#heading=h.rkdqb0ttlsiu).
+
+### Network Health Safeguards
+
+To ensure the hoarding system doesn't negatively impact network health:
+
+1. **Separation of Concerns**:
+   - Hoarding processes act as passive observers (do not forward messages)
+   - Full Cardano node handles all network relaying duties
+   - This maintains normal network citizenship without compromising data collection
+2. **Rate Limiting**: Implement connection and request rate limits to avoid overwhelming peers
+3. **Graceful Degradation**: Reduce activity if peers start rejecting connections
+4. **Transparent Observation**: Hoarding processes behave like normal peers (request blocks, stay synchronized) but don't relay data
+
+## Out of Scope
+
+This section clarifies architectural choices, features, and deliverables that are explicitly **not** part of this proposal to prevent scope creep and manage expectations.
```

Comment:
Almost exactly as you've described it, this is exactly the kind of material we are looking for in the _Rationale_ and should therefore be moved there... unless you want to make the case that the proposed nodes themselves are making such decisions about what data to collect (my apologies, but this is a quick triage review & I haven't been able to read for that kind of detail yet).